### PR TITLE
Add rule S4144 No identical functions

### LINKED ISSUE #15: Rule S4144: No identical functions
Implements [RSPEC-4144](https://jira.sonarsource.com/browse/RSPEC-4144)

### PR DESCRIPTION
**Pull request by saberduck**
fix #15

### CODE REVIEW

**Comment by saberduck** [comment]
Test is currently failing because I can't implemented exceptions the same way as they are done in Java. We should discuss how to handle this.

File: common-rule-engine/src/main/java/org/sonar/uast/helpers/FunctionLike.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  public static final UastNode.Kind KIND = UastNode.Kind.FUNCTION;
+  private final UastNode node;
+  private final UastNode block;
+
+  private FunctionLike(UastNode node, UastNode block) {
+    this.node = node;
+    this.block = block;
+  }
+
+  public static FunctionLike from(UastNode node) {
```
Comment (ghost) [inline common-rule-engine/src/main/java/org/sonar/uast/helpers/FunctionLike.java]: I would add a `@CheckForNull` annotation there.

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/NoIdenticalFunctionsCheck.java [hunk 2]
```diff
@@ -1,19 +1,40 @@
...
+    if (node.kinds.contains(UastNode.Kind.COMPILATION_UNIT) || node.kinds.contains(UastNode.Kind.CLASS)) {
+      functions.clear();
+    }
+    if (node.kinds.contains(FunctionLike.KIND)) {
+      FunctionLike thisFunction = FunctionLike.from(node);
+      if (thisFunction == null) {
+        return;
+      }
+      for (FunctionLike function : functions) {
+        if (syntacticallyEquivalent(thisFunction.body(), function.body())) {
```
Comment (ghost) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/NoIdenticalFunctionsCheck.java]: What if there is different arguments? Or empty bodies?

File: common-rule-engine/src/main/java/org/sonar/uast/helpers/FunctionLike.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package org.sonar.uast.helpers;
+
+import org.sonar.uast.UastNode;
+
+public class FunctionLike {
+
+  public static final UastNode.Kind KIND = UastNode.Kind.FUNCTION;
```
Comment (ghost) [inline common-rule-engine/src/main/java/org/sonar/uast/helpers/FunctionLike.java:9]: I would prefer `FUNCTION_KIND`

**Review by ghost — CHANGES_REQUESTED** [review]
@saberduck make the example compliant then. If we don't manage the same exceptions, let's at least accept them for the time being (so change the test file). I however thing that we could make the rule stricter and also covers the parameters.

File: common-rule-engine/src/main/java/org/sonar/uast/helpers/FunctionLike.java (hunk 3 above)
Comment (saberduck) [reply common-rule-engine/src/main/java/org/sonar/uast/helpers/FunctionLike.java:9]: I disagree, this constant should be used as `FunctionLike.KIND` , so having `FUNCTION` again in the name would be redundant

File: common-rule-engine/src/main/java/org/sonar/uast/helpers/FunctionLike.java (hunk 1 above)
Comment (saberduck) [reply common-rule-engine/src/main/java/org/sonar/uast/helpers/FunctionLike.java]: There should be a rule for this ;-)

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/NoIdenticalFunctionsCheck.java (hunk 2 above)
Comment (saberduck) [reply common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/NoIdenticalFunctionsCheck.java]: added comparison for parameters and exclude methods with less than 2 statements

**Comment by saberduck** [comment]
@m-g-sonar please have a look at my changes

**Review by ghost — APPROVED** [review]

**Pull request merged**